src/compiler/utilities.ts

```diff
@@ -6258,6 +6258,34 @@ namespace ts {
         return isSourceFile(node) || isModuleBlock(node) || isBlock(node) && isFunctionLike(node.parent);
     }
 
+    /* @internal */
+    export function isNewThisScope(node: Node): boolean {
+        switch (node.kind) {
+            case SyntaxKind.ArrowFunction:
+                switch (node.parent.kind) {
+                    case SyntaxKind.PropertyDeclaration:
+                        return true;
+                    default:
+                        return false;
+                }
+            case SyntaxKind.FunctionDeclaration:
+            case SyntaxKind.FunctionExpression:
+            case SyntaxKind.Parameter:

```

Comment:
this will falsely abort recursing into parameters of arrow functions

Reply:
Added a test case: `this` cannot be referenced in constructor arguments is the actual expected error there... which seems odd. I'll look into why and might file a separate issue.

Reply:
I created a PR to fix that error this week, don't know if it's already merged

Reply:
For posterity: https://github.com/Microsoft/TypeScript/pull/29298